beacon/light/committee_chain.go

```
@@ -0,0 +1,555 @@
+	s := &CommitteeChain{
+		fixedRoots: newCanonicalStore[common.Hash](db, fixedRootKey, func(root common.Hash) ([]byte, error) {
+			return root[:], nil
+		}, func(enc []byte) (root common.Hash, err error) {
+			if len(enc) == len(root) {
+				copy(root[:], enc)
+			} else {
+				err = errors.New("Incorrect length for committee root entry in the database")
+			}
+			return
+		}),
+		committees: newCanonicalStore[*types.SerializedSyncCommittee](db, syncCommitteeKey, func(committee *types.SerializedSyncCommittee) ([]byte, error) {
+			return committee[:], nil
+		}, func(enc []byte) (*types.SerializedSyncCommittee, error) {
+			if len(enc) == types.SerializedSyncCommitteeSize {
+				committee := new(types.SerializedSyncCommittee)
+				copy(committee[:], enc)
+				return committee, nil
+			}
+			return nil, errors.New("Incorrect length for serialized committee entry in the database")
+		}),
+		updates: newCanonicalStore[*types.LightClientUpdate](db, bestUpdateKey, func(update *types.LightClientUpdate) ([]byte, error) {
+			return rlp.EncodeToBytes(update)
+		}, func(enc []byte) (*types.LightClientUpdate, error) {
+			update := new(types.LightClientUpdate)
+			if err := rlp.DecodeBytes(enc, update); err != nil {
+				return nil, err
+			}
+			return update, nil
+		}),
+		syncCommitteeCache: lru.NewCache[uint64, syncCommittee](10),
+		db:                 db,
+		sigVerifier:        sigVerifier,
+		clock:              clock,
+		unixNano:           unixNano,
+		config:             config,
+		signerThreshold:    signerThreshold,
+		enforceTime:        enforceTime,
+		minimumUpdateScore: types.UpdateScore{
+			SignerCount:    uint32(signerThreshold),
+			SubPeriodIndex: params.SyncPeriodLength / 16,
+		},
+	}
+
+	// check validity constraints
+	if !s.updates.IsEmpty() {
+		if s.fixedRoots.IsEmpty() || s.updates.First < s.fixedRoots.First ||
+			s.updates.First >= s.fixedRoots.AfterLast {
+			log.Crit("Inconsistent database error: first update is not in the fixed roots range")
```

In general, I prefer things to return errors rather than `log.Crit`, since the latter doesn't permit graceful shutdown, and may cause additional 'damage'.

---

done